File: backend/app/core/apollo.py
import httpx

from app.core.config import settings
from app.schemas.enrichment import ApolloPersonResponse

APOLLO_BASE_URL = "https://api.apollo.io/api/v1"


def _headers() -> dict:
    return {
        "Content-Type": "application/json",
        "Cache-Control": "no-cache",
        "X-Api-Key": settings.apollo_api_key,
    }


def fetch_person_by_linkedin(linkedin_url: str) -> ApolloPersonResponse:
    response = httpx.post(
        f"{APOLLO_BASE_URL}/people/match",
        headers=_headers(),
        json={"linkedin_url": linkedin_url, "reveal_personal_emails": False},
        timeout=15,
    )
    response.raise_for_status()
    return ApolloPersonResponse(**response.json())


def fetch_person_by_email(email: str) -> ApolloPersonResponse:
    response = httpx.post(
        f"{APOLLO_BASE_URL}/people/match",
        headers=_headers(),
        json={"email": email, "reveal_personal_emails": False},
        timeout=15,
    )
    response.raise_for_status()
    return ApolloPersonResponse(**response.json())

File: backend/src/core/apollo.ts
import { Config } from "./config";
import { ApolloPersonResponseSchema, ApolloPersonResponse } from "../schemas/enrichment";
import { UpstreamError } from "./errors";

const APOLLO_BASE_URL = "https://api.apollo.io/api/v1";

export interface ApolloClient {
  fetchPersonByLinkedin(linkedinUrl: string): Promise<ApolloPersonResponse>;
  fetchPersonByEmail(email: string): Promise<ApolloPersonResponse>;
}

export function createApolloClient(config: Config): ApolloClient {
  async function post(body: Record<string, unknown>): Promise<ApolloPersonResponse> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 15_000);
    let res: Response;
    try {
      res = await fetch(`${APOLLO_BASE_URL}/people/match`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Cache-Control": "no-cache",
          "X-Api-Key": config.APOLLO_API_KEY,
        },
        body: JSON.stringify({ ...body, reveal_personal_emails: false }),
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timeout);
    }
    if (!res.ok) {
      throw new UpstreamError(`Apollo API error: ${res.status}`, res.status);
    }
    const json = await res.json();
    return ApolloPersonResponseSchema.parse(json);
  }

  return {
    fetchPersonByLinkedin: (linkedin_url) => post({ linkedin_url }),
    fetchPersonByEmail: (email) => post({ email }),
  };
}
